Feature/chilgenb add to back tracker

chilge opened the pull request:
Fixes a bug arising from the feature that a single sdp::EnergyDeposit can be associated with several raw::RawDigits, leading to multiple counting . This is done by adding the weights, or fraction of a given energy deposit's energy, assigned by the pad response functions (PRFs) to the art::Assns.

chilge committed:
add method to retrieve total ionization energy associated with Track

chilge committed:
add method to retrieve total ionization energy associated with Track

chilge committed:
move HeaderTree fill to endJob

chilge committed:
Merge remote-tracking branch 'origin/develop' into feature/chilgenb_AddToBackTracker

chilge committed:
Merge branch 'develop' into feature/chilgenb_AddToBackTracker

chilge committed:
Merge branch 'develop' into feature/chilgenb_AddToBackTracker

chilge committed:
fix bug in < operator and add == operator

chilge committed:
new analyzer for testing the backtracker

chilge committed:
new methods for getting total track energy and backtracking for TPCCluster

chilge committed:
add missing process names

chilge committed:
add new defs for associations with float in data field instead of void

chilge committed:
add new method for making associations with non-void data field

chilge committed:
add missing lines for ECal properties

chilge committed:
additions for writing fraction of sdp::EnergyDeposit energy associated with RawDigit to assns

chilge committed:
updates to accomodate fetching data field from RawDigit-EnergyDeposit art::assns for splitting up energy deposits, made FHiCL configurable (enabled by default)

chilge committed:
add a few more histograms

chilge committed:
conversion of README.rtf since .md is GitHub friendly

chilge committed:
pull in changes from devlop and fix merge conflicts

chilge committed:
Update README.md

fix some formatting issues left over from the initial rtf conversion; add documentation on new method (TrackToTotalEnergy)

bellanto commented:
Looks mostly good to me, but I haven’t tried to build or run or debug it.  Three concerns, and the rest are nits to pick.


My first concern is with the == and < operators for HitIDE.  The latter has 2x as many lines as it needs... but I’m more concerned with what are the appropriate sorting criteria.  They should be the same for both operators.  The comment says you sort on time, but there isn’t a time.  And there are equality tests for real numbers, that’s not thick ice.  I guess I’d sort on TrackID and then energyTot, or maybe hitr.position.Z()... I’m not sure because I haven’t quite gotten the use cases worked out for these operators.  And then, whichever you decide, it should be the same for CalIDE.  Well, ok, ya can’t cut on hitr.position.Z() in CalIDE, right?

My second concern is why the std::vector<const rec::TPCCluster*> const TrackToClusters signature and not the std::vector<art::Ptr<recTPCCluster*>> signature?  All the similar methods return art::Ptrs.  And the same for the associated structure fTrackIDToClusters.

My third concern is that at line 213 of Backtracker_service.cc, when constructing fChannelToEDepCol, you replaced the old swap(eDeps) with an emplace_back(blah blah blah) and it just sets off an alarm in my head.  I’m trying to remember why we had the swap in there.  Does is delete the existing data?  Is that it?  Or am I just not remembering correctly?  It might not matter at all.

Here are the nits:

         1) Line 303 in BackTrackerCore.cxx: Did you really see negative start or stop times?  That should not happen and if it does that’s a real bug.
         2) Similarly, at the end of HitToHitIDEs in BackTrackerCore.cxx: Did you really see multiple IDEs in a single channel?
         3) I suggest replacing in BackTrackerCore::TrackTotalEnergy

      /*size_t pretotal = allIdes.size();

      // remove duplicate HitIdes
      // comment out for now since we could have meaningful duplicate HitIDEs
      // e.g. a common sdp::EnergyDeposit split evenly among channels
      std::sort(allIdes.begin(),allIdes.end());
      allIdes.erase( std::unique(allIdes.begin(),allIdes.end()), allIdes.end());
      size_t posttotal = allIdes.size();

      if(pretotal!=posttotal) std::cout << "removed " << pretotal-posttotal << " duplicated IDEs" << std::endl;*/


With the substantially more succinct

     // Don’t check for duplicate HitIDEs as they can occur when an sdp::EnergyDeposit is split between channels

         4) Please use 4 spaces for indentation like the rest of the code.  And no tabs.  Tabs are like arcane prophecies from the Oracles at Delphi.  Every editor interprets them as it chooses.  Evil!  Evil!  Actually it looks like you killed some indent problems already.
         5) Please stick 4 more carriage returns between lines 162 164 of BackTrackerCore.h, just after ParticleToMCTruth(simb::MCParticle* const p) const.  This is like the space after HitEfficiency(blah blah blah).

chilge committed:
fix merge error (inconsistency in capitalization)

chilge committed:
pick up needed changes from garana

chilge committed:
seems that art got pickier

chilge committed:
Merge branch 'develop' into feature/chilgenb_AddToBackTracker

chilge commented:
Thank you, @bellanto, for the thorough review!

Concerns
1. I actually think that the number of if statements present is required to get a unique sorting. I don't understand what you mean by sorting criteria for the == operator where the requirement is that all fields are equal. For what concerns the sorting order for the < operator, the only use case I am aware of at the moment is for applying std::search type algorithms. If that is the case, the sort order doesn't really matter. I do agree that CalIDE needs equivalent operators and will implement that. I added the missing check on the time stamp.

2. Good catch. I swapped that method over to using art::Ptr.

3. I don't believe that using emplace_back introduces any issue. the map gets cleared before filling so each map value is just an empty vector once the keys are declared. The swap just moves values from one container to another. I believe these two implementations are functionally identical. I could be wrong, but I think my implementation is slightly more efficient.

Nits
1. No I don't. I added that line when I was debugging and decided to keep it in the end, just in case.
2. No I don't. Again, just kept it there as a sanity check.
3. Fair enough. Done.
4. Preaching to the choir about tabs. Just did a search and replace using \t and I only found 3 instances. I believe most of the 2 space/4 space inconsistency was present before I modified anything. I went thorough and tried to enforce consistency throughout. 
5. What is the reasoning? I did it, just curious.
